cli_tools/google-osconfig-agent/patch/patch_run.go

```diff
@@ -0,0 +1,161 @@
+//  Copyright 2018 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package patch
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/cloud.google.com/go/osconfig/apiv1alpha1"
+	osconfigpb "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha1"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/config"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/logger"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/tasker"
+	"github.com/golang/protobuf/jsonpb"
+	"google.golang.org/api/option"
+)
+
+// Init starts the patch system.
+func Init(ctx context.Context) {
+	// Load current patch state off disk.
+	pr, err := loadState(state)
+	if err != nil {
+		logger.Errorf("loadState error: %v", err)
+	} else if pr != nil && !pr.Complete {
+		client, err := osconfig.NewClient(ctx, option.WithEndpoint(config.SvcEndpoint()), option.WithCredentialsFile(config.OAuthPath()))
+		if err != nil {
+			logger.Errorf("osconfig.NewClient Error: %v", err)
+		} else {
+			tasker.Enqueue("Run patch", func() { patchRunner(ctx, client, pr) })
+		}
+	}
+
+	go watcher(ctx)
+}
+
+type patchRun struct {
```

Comment:
It's a mouthful but the service calls this: InstancePatchStatus and it includes the current step/state of the instance's patch application. If these are the same, and I'm not entirely sure they are, it would be best to be consistent with the naming.